diff --git a/examples/LIBERO/finetune_libero_10.sh b/examples/LIBERO/finetune_libero_10.sh
index cbaea8f..7ecd313 100644
--- a/examples/LIBERO/finetune_libero_10.sh
+++ b/examples/LIBERO/finetune_libero_10.sh
@@ -15,7 +15,6 @@ torchrun --nproc_per_node=$NUM_GPUS --master_port=29500 \
     --warmup_ratio 0.05 \
     --weight_decay 1e-5 \
     --learning_rate 1e-4 \
-    --use_wandb \
     --global_batch_size 640 \
     --color_jitter_params brightness 0.3 contrast 0.4 saturation 0.5 hue 0.08 \
     --dataloader_num_workers 4 \
diff --git a/gr00t/configs/training/training_config.py b/gr00t/configs/training/training_config.py
index da488eb..106638d 100755
--- a/gr00t/configs/training/training_config.py
+++ b/gr00t/configs/training/training_config.py
@@ -32,7 +32,7 @@ class TrainingConfig:
     start_from_checkpoint: Optional[str] = None
 
     # Mixed precision
-    tf32: bool = True
+    tf32: bool = False
     fp16: bool = False
     bf16: bool = True
     eval_bf16: bool = True
diff --git a/gr00t/experiment/launch_finetune.py b/gr00t/experiment/launch_finetune.py
index f5deb7e..1f934c8 100644
--- a/gr00t/experiment/launch_finetune.py
+++ b/gr00t/experiment/launch_finetune.py
@@ -9,7 +9,10 @@ import tyro
 from gr00t.configs.base_config import get_default_config
 from gr00t.configs.finetune_config import FinetuneConfig
 from gr00t.experiment.experiment import run
-
+import torch
+import torch_npu
+from torch_npu.contrib import transfer_to_npu
+from gr00t.utils.patch import generate_patcher_builder
 
 # Make sure the user provided modality config is registered.
 def load_modality_config(modality_config_path: str):
@@ -26,6 +29,7 @@ def load_modality_config(modality_config_path: str):
 
 
 if __name__ == "__main__":
+    generate_patcher_builder()
     # Set LOGURU_LEVEL environment variable if not already set (default: INFO)
     if "LOGURU_LEVEL" not in os.environ:
         os.environ["LOGURU_LEVEL"] = "INFO"
diff --git a/gr00t/model/modules/embodiment_conditioned_mlp.py b/gr00t/model/modules/embodiment_conditioned_mlp.py
index d477898..2eef027 100644
--- a/gr00t/model/modules/embodiment_conditioned_mlp.py
+++ b/gr00t/model/modules/embodiment_conditioned_mlp.py
@@ -48,6 +48,8 @@ class CategorySpecificLinear(nn.Module):
         super().__init__()
         self.num_categories = num_categories
         # For each category, we have separate weights and biases.
+        self.input_dim = input_dim
+        self.hidden_dim = hidden_dim
         self.W = nn.Parameter(0.02 * torch.randn(num_categories, input_dim, hidden_dim))
         self.b = nn.Parameter(torch.zeros(num_categories, hidden_dim))
 
@@ -59,8 +61,12 @@ class CategorySpecificLinear(nn.Module):
         Returns:
             [B, T, hidden_dim] output tensor
         """
-        selected_W = self.W[cat_ids]
-        selected_b = self.b[cat_ids]
+        B = cat_ids.shape[0]
+        one_hot = F.one_hot(cat_ids, num_classes=self.num_categories).to(dtype=x.dtype, device=x.device)
+        W_flat = self.W.view(self.num_categories, -1)
+        selected_W_flat = torch.matmul(one_hot, W_flat)  # [B, input_dim*hidden_dim]
+        selected_W = selected_W_flat.view(B, self.input_dim, self.hidden_dim)
+        selected_b = torch.matmul(one_hot, self.b)
         return torch.bmm(x, selected_W) + selected_b.unsqueeze(1)
 
     def expand_action_dimension(
diff --git a/gr00t/model/modules/nvidia/Eagle-Block2A-2B-v2/modeling_siglip2.py b/gr00t/model/modules/nvidia/Eagle-Block2A-2B-v2/modeling_siglip2.py
index 94251a4..a363c22 100644
--- a/gr00t/model/modules/nvidia/Eagle-Block2A-2B-v2/modeling_siglip2.py
+++ b/gr00t/model/modules/nvidia/Eagle-Block2A-2B-v2/modeling_siglip2.py
@@ -56,7 +56,7 @@ from typing import Optional, Tuple
 import torch
 import torch.nn.functional as F
 
-from transformers.utils import is_flash_attn_2_available, is_flash_attn_greater_or_equal, logging
+from transformers.utils import is_flash_attn_2_available, is_flash_attn_greater_or_equal, logging, is_torch_npu_available
 from transformers.integrations.flash_attention import flash_attention_forward as original_flash_attention_forward
 flash_241 = is_flash_attn_greater_or_equal("2.4.1")
 deterministic_g = os.environ.get("FLASH_ATTENTION_DETERMINISTIC", "0") == "1"
@@ -71,6 +71,11 @@ if is_flash_attn_2_available():
 
     _flash_supports_window_size = "window_size" in list(inspect.signature(flash_attn_func).parameters)
 
+if is_torch_npu_available():
+    from transformers.integrations.npu_flash_attention import index_first_axis, pad_input, unpad_input
+    from transformers.integrations.npu_flash_attention import npu_flash_attn_func as flash_attn_func
+    from transformers.integrations.npu_flash_attention import npu_flash_attn_varlen_func as flash_attn_varlen_func
+    _flash_supports_window_size = "window_size" in list(inspect.signature(flash_attn_func).parameters)
 
 def _flash_attention_forward(
     query_states: torch.Tensor,
diff --git a/pyproject.toml b/pyproject.toml
index 93eb2e0..8c8d0c1 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -24,7 +24,6 @@ dependencies = [
     "torchvision==0.22.1",
     "transformers==4.51.3",
     "tyro==0.9.17",
-    "flash-attn==2.7.4.post1",
     "click==8.1.8",
     "datasets==3.6.0",
     "einops==0.8.1",
@@ -33,7 +32,6 @@ dependencies = [
     "numpy==1.26.4",
     "omegaconf==2.3.0",
     "scipy==1.15.3",
-    "torchcodec==0.4.0",
     "wandb==0.23.0",
     "pyzmq==27.0.1",
     "deepspeed==0.17.6",
@@ -53,8 +51,6 @@ tensorrt = [
 where = ["."]
 include = ["gr00t*"]
 
-[tool.uv.extra-build-dependencies]
-flash-attn = ["torch==2.7.0", "numpy==1.26.4"]
 
 [tool.ruff]
 line-length = 100
